core/src/test/java/org/apache/calcite/rex/RexExecutorTest.java

```diff
@@ -378,4 +383,140 @@ public void run() {
   interface Action {
     void check(RexBuilder rexBuilder, RexExecutorImpl executor);
   }
+
+  /**
+   * ArrayList-based DataContext to check Rex execution.
+   */
+  public static class TestDataContext extends SingleValueDataContext {
+    private TestDataContext(Object[] values) {
+      super("inputRecord", values);
+    }
+  }
+
+  /**
+   * Context that holds a value for a particular context name.
+   */
+  static class SingleValueDataContext implements DataContext {
+    private final String name;
+    private final Object value;
+
+    SingleValueDataContext(String name, Object value) {
+      this.name = name;
+      this.value = value;
+    }
+
+    public SchemaPlus getRootSchema() {
+      throw new RuntimeException("Unsupported");
+    }
+
+    public JavaTypeFactory getTypeFactory() {
+      throw new RuntimeException("Unsupported");
+    }
+
+    public QueryProvider getQueryProvider() {
+      throw new RuntimeException("Unsupported");
+    }
+
+    public Object get(String name) {
+      if (this.name.equals(name)) {
+        return value;
+      } else {
+        throw new RuntimeException("Wrong DataContext access");
+      }
+    }
+  }
+
+  /**
+   * User defined executable.
+   */
+  static class UserExecutable implements RexExecutable {
+    final RexBuilder rexBuilder;
+    final List<RexNode> exps;
+    DataContext dataContext;
+    UserExecutable(final RexBuilder rexBuilder, final List<RexNode> exps) {
```

Comment:
DataContext dataContext -> final  DataContext dataContext ?

Reply:
setDataContext will change the dataContext